compiler/rustc_data_structures/src/stack.rs

```diff
@@ -13,5 +13,32 @@ const STACK_PER_RECURSION: usize = 1 * 1024 * 1024; // 1MB
 ///
 /// Should not be sprinkled around carelessly, as it causes a little bit of overhead.
 pub fn ensure_sufficient_stack<R>(f: impl FnOnce() -> R) -> R {
-    stacker::maybe_grow(RED_ZONE, STACK_PER_RECURSION, f)
+    let mut mut_closure = Some(f);
+    let mut retval: Option<R> = None;
+    let retval_ref = &mut retval;
+
+    // This wrapper around `f` achieves two things:
+    // * It converts the `impl FnOnce` to a `dyn FnMut`.
+    //   `dyn` because we want it to not be generic, and
+    //   `FnMut` because we can't pass a `dyn FnOnce` around without boxing it.
+    // * It eliminates the generic return value, by writing it to the stack of this function.
+    //   Otherwise the closure would have to return an unsized value, which isn't possible.
+    let dyn_closure: &mut dyn FnMut() = &mut || {
+        // The closure needs to be owned to be invoked, and with this trick
+        // we "move" it into this borrowed closure.
+        let closure = mut_closure.take().unwrap();
+        *retval_ref = Some(closure());
+    };
+
+    nongeneric_ensure_sufficient_stack(dyn_closure);
+    retval.unwrap()
+}
+
+/// This optimises the compile time of rustc:
+/// Measuring with cargo-llvm-lines revealed that `psm::on_stack::with_on_stack` was
+/// monomorphized 1552 times and was responsible for 1.5% of rustc's total llvm-lines.
+/// Making this wrapper without a generic bound removes all of that duplication.
+#[inline(never)]
+fn nongeneric_ensure_sufficient_stack(f: &mut dyn FnMut()) {
+    stacker::maybe_grow(RED_ZONE, STACK_PER_RECURSION, || f());
```

Comment:
You can avoid the extra closure:
```suggestion
    stacker::maybe_grow(RED_ZONE, STACK_PER_RECURSION, f);
```

Reply:
Neat. Didn't know this would work here.